core/src/main/java/org/apache/iceberg/CachingCatalog.java

```diff
@@ -20,33 +20,144 @@
 import org.apache.iceberg.catalog.TableIdentifier;
 import org.apache.iceberg.exceptions.AlreadyExistsException;
 import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class CachingCatalog implements Catalog {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CachingCatalog.class);
+  private static final RemovalListener<TableIdentifier, Table> keyLoggingRemovalListener =
+      (key, value, cause) -> LOG.info("Expired {} from the TableCache", key);
+
   public static Catalog wrap(Catalog catalog) {
-    return wrap(catalog, true);
+    return wrap(catalog, false, 0);
+  }
+
+  public static Catalog wrap(Catalog catalog, boolean expirationEnabled, long expirationIntervalMilllis) {
+    return wrap(catalog, true, expirationEnabled, expirationIntervalMilllis);
+  }
+
+  public static Catalog wrap(Catalog catalog, boolean caseSensitive, boolean expirationEnabled,
+      long expirationIntervalMillis) {
+    return new CachingCatalog(catalog, caseSensitive, expirationEnabled, expirationIntervalMillis);
   }
 
-  public static Catalog wrap(Catalog catalog, boolean caseSensitive) {
-    return new CachingCatalog(catalog, caseSensitive);
+  public static Catalog wrap(Catalog catalog, boolean expirationEnabled, long expirationIntervalMillis, Ticker ticker) {
+    return new CachingCatalog(catalog, true, expirationEnabled, expirationIntervalMillis, ticker);
   }
 
-  private final Cache<TableIdentifier, Table> tableCache = Caffeine.newBuilder().softValues().build();
   private final Catalog catalog;
   private final boolean caseSensitive;
+  private final boolean expirationEnabled;
+  private final long expirationIntervalMillis;
+  private Cache<TableIdentifier, Table> tableCache;
 
-  private CachingCatalog(Catalog catalog, boolean caseSensitive) {
+  private CachingCatalog(Catalog catalog, boolean caseSensitive, boolean isExpirationEnabled,
+      long expirationIntervalInMillis) {
+    this(catalog, caseSensitive, isExpirationEnabled, expirationIntervalInMillis, Ticker.systemTicker());
+  }
+
+  private CachingCatalog(Catalog catalog, boolean caseSensitive, boolean isExpirationEnabled,
+      long expirationIntervalMillis, Ticker ticker) {
     this.catalog = catalog;
     this.caseSensitive = caseSensitive;
+    this.expirationEnabled = isExpirationEnabled;
+    this.expirationIntervalMillis = expirationIntervalMillis;
+
+    this.tableCache = createTableCache(ticker);
+  }
+
+  /**
+   * Return the age of an entry in the cache.
+   * <p>
+   * This method is only visiable for testing the cache expiration policy, as cache invalidation is handled
+   * by the catalog and not the cache itself.
+   * <p>
+   * Returns the age of the cache entry corresponding to the identifier,  or {@code Optional.empty} if the table
+   * is not present in the cache or if no expireAfterAccess policy is present in this CachingCatalog.
+   */
+  // VisibleForTesting
```

Comment:
This unfortunately can't use the @VisibleForTesting annotation, as the checkstyle complains that it's not package private (the tests are in org.apache.iceberg.hadoop and not org.apache.iceberg).

I was told I can use a helper in the tests from within the same package to have it be `@VisibleForTesting` again, so will try that.